### PR TITLE
Relative lookup of template-hsc.h

### PR DESCRIPTION
**Pull request by angerman**
As a last resort, try to find the `template-hsc.h` in `../lib` relative to the hsc2hs executable.

### CODE REVIEW

**Comment by hvr** [comment]
Tbh, this looks suspicious. `hsc2hs` should rather become independent from the compiler it was built with. This seems to go in the opposite direction I envisioned for `hsc2hs`.

**Comment by hvr** [comment]
What's wrong with the current approach of using Cabal's `getPackageData` facility to locate `hsc2hs`'s hsc template? After all, we must be sure to pick the template that came with the `hsc2hs` release, not the one that happens to be bundled with GHC, which can be either outdated or too new for currently invoked `hsc2hs` version.

**Comment by hvr** [comment]
Update: we discussed this a bit off-thread; the motivation is mostly to support relocatable GHC bindists; and the short-term hack (ideally, Cabal would provide us a more principled way to support relocatable programs & packages)  is to hide this logic behind a manual cabal flag that's off by default, and turned on by the GHC buildsystem (similar to how e.g. `haddock` does this)

File: hsc2hs.cabal [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Name: hsc2hs
-Version: 0.68.2
+Version: 0.69.0
```
Comment (hvr) [inline hsc2hs.cabal]: Why this major version bump? Note that it's recommended to write dependencies such as  `build-tools: hsc2hs >= 0.68.2 && < 0.69`; major version bumps in `hsc2hs` should be reserved when its generated output changes in an incompatible way; but it doesn't here.

File: Main.hs [hunk 2]
```diff
@@ -13,21 +13,39 @@ import Control.Monad            ( liftM, forM_ )
...
+-- If we ware building the hsc2hs
+-- binary for binary distribution
+-- in the GHC tree.  And if ghc-boot
+-- is 8.3 or newer, we can use the
+-- @getBaseDir@ function to obtain
+-- the path to the @$topdir/lib@
+-- folder, and try to locate the
+-- @template-hsc.h@ there.
+#if defined(IN_GHC_TREE)
+# if defined(MIN_VERSION_ghc_boot)
```
Comment (hvr) [inline Main.hs]: I'd do this differently in a "more defined" way; if `IN_GHC_TREE` is defined, we have already ensured via the `.cabal` file that ghc-boot >= 8.3; IOW I'd do this:

```c
#if defined(IN_GHC_TREE) 
# if !MIN_VERSION_ghc_boot(8,3,0)
#  error the impossible happened
# endif
#endif
```

and then we can avoid `#define`ing yet another CPP macro...

File: hsc2hs.cabal [hunk 3]
```diff
@@ -50,4 +55,10 @@ Executable hsc2hs
                    directory  >= 1   && < 1.4,
                    filepath   >= 1   && < 1.5,
                    process    >= 1.1 && < 1.7
+    if flag(in-ghc-tree)
+       if impl(ghc>=8)
+          cpp-options: -DIN_GHC_TREE
+          -- use @getBaseDir@ from ghc-boot to locate
+          -- the template-hsc.h file.
+          build-depends: ghc-boot
```
Comment (hvr) [inline hsc2hs.cabal]: ...and then this should rather read something like `build-depends: ghc-boot >= 8.3 && < 8.5`

File: Main.hs [hunk 4]
```diff
@@ -13,21 +13,39 @@ import Control.Monad            ( liftM, forM_ )
...
 import System.IO
 
 #ifdef BUILD_NHC
 import System.Directory         ( getCurrentDirectory )
 #else
-import Data.Version             ( showVersion )
-import Paths_hsc2hs as Main     ( getDataFileName, version )
+import Paths_hsc2hs as Main     ( getDataFileName )
+#endif
+#if defined(HAVE_BasePath)
```
Comment (hvr) [inline Main.hs]: ...and here we can say `#if defined(IN_GHC_TREE)` instead...

File: Main.hs [hunk 5]
```diff
@@ -49,12 +67,10 @@ import HSCParser
 #ifdef BUILD_NHC
 getDataFileName s = do here <- getCurrentDirectory
                        return (here++"/"++s)
-version = "0.67" -- TODO!!!
-showVersion = id
 #endif
 
 versionString :: String
-versionString = "hsc2hs version " ++ showVersion version ++ "\n"
+versionString = "hsc2hs version " ++ CURRENT_PACKAGE_VERSION ++ "\n"
```
Comment (hvr) [inline Main.hs]: (why this change btw?)

File: Main.hs [hunk 6]
```diff
@@ -141,6 +157,9 @@ findTemplate usage mb_libdir config
      --
      -- Next we try the location we told Cabal about.
      --
+     -- Finally we also try to locate the template in the `baseDir`
+     -- as provided by the `ghc-boot` library.
```
Comment (hvr) [inline Main.hs]: ...comment should clarify this is only done as a hack when `IN_GHC_TREE` is defined...

File: Main.hs [hunk 7]
```diff
@@ -157,14 +176,27 @@ findTemplate usage mb_libdir config
...
              templ2 <- getDataFileName "template-hsc.h"
              exists2 <- doesFileExist templ2
-             if exists2 then return (templ2, [])
-                        else die ("No template specified, and template-hsc.h not located.\n\n" ++ usage)
+             if exists2
+                then return $ Just (templ2, [])
+                else return Nothing
+     case mb_templ2 of
+         Just x -> return x
+#if defined(HAVE_BasePath)
```
Comment (hvr) [inline Main.hs]: ...and use `#if defined(HAVE_BasePath)` instead...

File: hsc2hs.cabal (hunk 3 above)
Comment (angerman) [reply hsc2hs.cabal]: Haddock (and other packages depending on ghc-boot) also have no version bounds, unless they are `== ` and set via autoconf to the munged project version.

File: Main.hs (hunk 5 above)
Comment (angerman) [reply Main.hs]: `showVersion` and `version` seemed to be relicts as `CURRENT_PACKAGE_VERSION` was now available? Am I mistaken?

Comment (hvr) [reply Main.hs]: I should have paid more attention... see #6 :-)
